### PR TITLE
Offers comments

### PR DESCRIPTION
**Pull request by jonzee**
Implemented `Comments` component with external data & pagination. Styles was improved in #24 so I didn't touch anything (it'll be ok after merge).

### CODE REVIEW

File: src/ts/Detail/Comments.spec.tsx [hunk 1]
```diff
@@ -2,17 +2,21 @@ import 'mocha';
 import * as React from 'react';
 import { expect } from 'chai';
 import Comments from './Comments';
-import { shallow } from 'enzyme';
+import { mountWithTheme } from '../helpers/test-theme';
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.spec.tsx]: 😍

File: src/ts/Detail/Comments.spec.tsx [hunk 2]
```diff
@@ -2,17 +2,21 @@ import 'mocha';
...
 
 describe('<Comments />', () => {
-  it('check for inner components: comments, rate, author', () => {
-    const wrapper = shallow(<Comments />);
+  it('check for inner components: comments, rate, author', function(done) {
+    this.timeout(3000);
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.spec.tsx]: we cannot have non-zero timeouts. They will significantly slow the tests down, and are unreliable. Let's use resolves instead.

File: src/ts/Detail/Comments.tsx [hunk 3]
```diff
@@ -4,66 +4,130 @@ import {
...
+    }
+    this.comment = this.comment.bind(this);
+    this.rating = this.rating.bind(this);
+    this.fetch = this.fetch.bind(this);
+  }
+
+  componentDidMount() { this.fetch(); }
+
+  fetch() {
+    this.setState({ fetching: true });
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: almost like redux :+1:

File: src/ts/Detail/Comments.tsx [hunk 4]
```diff
@@ -4,66 +4,130 @@ import {
...
+    }
+
+    const rating = this.rating();
+    const comments = this.comment();
+    const more = this.state.fetching ? <div>Loading comments</div> : <div onClick={this.fetch}>View all references</div>;
+    return (
+      <div className="comments">
+        <Paper style={{ width: '100%' }} zDepth={2} rounded={false}>
+          <div className="comments-cont">
+            <div className="references">{this.state.references} References</div>
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: the two returns seem only to differ with the contents of `.comments-cont` – let's keep it DRY

File: src/ts/Detail/Comments.tsx [hunk 5]
```diff
@@ -4,66 +4,130 @@ import {
...
+    }, 2000);
+  }
+
+  render() {
+    if (this.state.loading) {
+      return (
+        <div className="comments">
+          <Paper style={{ width: '100%' }} zDepth={2} rounded={false}>
+            <div className="comments-cont">
+              <div style={{textAlign: 'center'}}>LOADING COMMENTS...</div>
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: why so inline css

File: src/ts/Detail/Comments.spec.tsx (hunk 2 above)
Comment (jonzee) [inline src/ts/Detail/Comments.spec.tsx]: It's only temporary for `API` mockup where I used timeout to simulate request.

File: src/ts/Detail/Comments.tsx [hunk 6]
```diff
@@ -4,66 +4,130 @@ import {
...
-              </div>
-              <div className="rate">
-                <div className="type">Rating</div>
-                <div className="stars">{stars}</div>
-              </div>
+export default class Comments extends React.Component<any, any> {
+  constructor(props) {
+    super(props);
+    this.state = {
+      offset: 0,
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: http://cdn.smosh.com/sites/default/files/bloguploads/u-mad-1.jpg

File: src/ts/Detail/Comments.spec.tsx [hunk 7]
```diff
@@ -5,18 +5,20 @@ import Comments from './Comments';
 import { mountWithTheme } from '../helpers/test-theme';
 
 describe('<Comments />', () => {
-  it('check for inner components: comments, rate, author', function(done) {
+  it('check for inner components: comments, rate, author', function (done) {
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.spec.tsx]: ok

File: src/ts/Detail/Comments.spec.tsx [hunk 8]
```diff
@@ -5,18 +5,20 @@ import Comments from './Comments';
...
+      () => {
+        expect(wrapper.find('.view-all')).to.have.length(1);
+        expect(wrapper.find('.type')).to.have.length(4);
+        const author = wrapper.find('.author').first();
+        expect(author.find('.picture')).to.have.length(1);
+        expect(author.find('.fullname')).to.have.length(1);
+        expect(author.find('.location')).to.have.length(1);
+        done();
+      },
+      2500);
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.spec.tsx]: no.

File: src/ts/Detail/Comments.tsx [hunk 9]
```diff
@@ -25,47 +25,43 @@ export default class Comments extends React.Component<any, any> {
...
-        rating: comments.rating,
-        references: comments.references,
-        comments: [
-          ...this.state.comments,
-          ...comments.comments,
-        ],
-      });
-    }, 2000);
+    setTimeout(
+      () => {
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: previous line

File: src/ts/Detail/Comments.tsx [hunk 10]
```diff
@@ -25,47 +25,43 @@ export default class Comments extends React.Component<any, any> {
...
+        this.setState({
+          loading: false,
+          fetching: false,
+          rating: comments.rating,
+          references: comments.references,
+          comments: [
+            ...this.state.comments,
+            ...comments.comments,
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: ???

File: src/ts/Detail/Comments.tsx [hunk 11]
```diff
@@ -25,47 +25,43 @@ export default class Comments extends React.Component<any, any> {
...
+          ],
+        });
+      },
+      2000);
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: no

File: src/ts/Detail/Comments.tsx [hunk 12]
```diff
@@ -25,47 +25,43 @@ export default class Comments extends React.Component<any, any> {
...
   }
 
   render() {
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: good with the deletion 👍

File: src/ts/Detail/Comments.tsx [hunk 13]
```diff
@@ -82,11 +78,11 @@ export default class Comments extends React.Component<any, any> {
           </div>
           <div className="text">{text}</div>
         </div>);
-      })
+      });
   }
 
   stars(rate) {
-    return _.times(5, i => {
+    return _.times(5, (i) => {
```
Comment (michalmikolajczyk) [inline src/ts/Detail/Comments.tsx]: why the parentheses for the sole parameter?

File: src/ts/Detail/Comments.tsx (hunk 6 above)
Comment (jonzee) [inline src/ts/Detail/Comments.tsx]: nah it was before our `offset` vs. `page`conversation

File: src/ts/Detail/Comments.tsx (hunk 10 above)
Comment (jonzee) [inline src/ts/Detail/Comments.tsx]: http://vincent.billey.me/pure-javascript-immutable-array

File: src/ts/Detail/Comments.tsx (hunk 11 above)
Comment (jonzee) [inline src/ts/Detail/Comments.tsx]: it's because of linter `arguments are not aligned`

File: src/ts/Detail/Comments.tsx (hunk 13 above)
Comment (jonzee) [inline src/ts/Detail/Comments.tsx]: airbnb style guide

**Pull request merged**